lib/component.js

```diff
@@ -652,6 +676,68 @@ class Component extends WebComponent {
       }
     }
   }
+
+  _findNearestPanelAncestor() {
+    if (!this.isConnected) {
+      throw new Error(`Cannot determine context before component is connected to the DOM`);
+    }
+
+    let node = this.parentNode;
+    while (node) {
+      if (node._getActiveContexts) {
+        return node;
+      }
+      if (node.constructor.name === `ShadowRoot`) {
+        node = node.host;
+      } else {
+        node = node.parentNode;
+      }
+    }
+    return null;
+  }
+
+  _findAndMergeActiveContexts() {
+    const contextParent = this._findNearestPanelAncestor();
+    const defaultContexts = this.getConfig(`defaultContexts`);
+
+    if (contextParent) {
+      const parentContexts = contextParent._getActiveContexts();
+
+      if (parentContexts.size > 0) {
+        // merge context maps
+        const combinedContexts = new Map(parentContexts.entries());
+        for (const [baseContextClass, defaultContext] of defaultContexts.entries()) {
+          if (!combinedContexts.has(baseContextClass)) {
+            combinedContexts.set(baseContextClass, defaultContext);
+          }
+        }
+
+        return combinedContexts;
+      }
+    }
+
+    return defaultContexts;
+  }
+
+  _getActiveContexts() {
+    if (!this._cachedActiveContexts) {
+      this._cachedActiveContexts = this._findAndMergeActiveContexts();
+    }
+    return this._cachedActiveContexts;
+  }
+
+  getContext(contextBaseClass) {
+    if (!contextBaseClass) {
+      throw new Error(`@contextBaseClass is null or undefined`);
+    }
+    const defaultContexts = this.getConfig(`defaultContexts`);
+    const defaultContext = defaultContexts.get(contextBaseClass);
+    if (!defaultContext) {
+      throw new Error(`A default context must be declared in config.defaultContexts before this context can be used.`);
+    }
+
+    return this._getActiveContexts().get(contextBaseClass);
```

Comment:
Cool. So we only start looking up the ancestor tree only when something calls `getContext`, and then we cache the contexts once they are resolved